### PR TITLE
swev-id: pytest-dev__pytest-7432 fix: preserve skip item location under --runxfail

### LINKED ISSUE #29: skipping: --runxfail breaks pytest.mark.skip location reporting
Issue: skipping: --runxfail breaks pytest.mark.skip location reporting
Pytest versions: 5.4.x, current master

When @pytest.mark.skip/skipif marks are used to skip a test, for example:

```py
import pytest
@pytest.mark.skip
def test_skip_location() -> None:
    assert 0
```

The expected skip location reported should point to the item itself, and this is indeed what happens when running with `pytest -rs`:

```
SKIPPED [1] test_it.py:3: unconditional skip
```

However, adding `pytest -rs --runxfail` breaks this:

```
SKIPPED [1] src/_pytest/skipping.py:238: unconditional skip
```

The `--runxfail` is only about xfail and should not affect this at all.

---

Research specification (by Emerson Gray):

Root cause:
- In `src/_pytest/skipping.py`, within `pytest_runtest_makereport`, there is an unconditional short-circuit when `--runxfail` is active:
  - `elif item.config.option.runxfail: pass  # don't interfere`
- This prevents the subsequent skip-location rewrite block from executing, leaving `rep.longrepr` pointing to where `skip()` is raised in the plugin (skipping.py), instead of the test item’s location.
- Without `--runxfail`, the skip-location rewrite runs and correctly reports the item’s file:line.

Proposed fix (minimal and correct):
- Remove the broad `runxfail` short-circuit so the skip-location rewrite can still run when `--runxfail` is active.
- Guard only the xfail-specific branches with `not item.config.option.runxfail` to preserve `--runxfail` semantics for xfail while leaving skip handling untouched:
  - Change `elif call.excinfo and isinstance(call.excinfo.value, xfail.Exception):` to `elif (not item.config.option.runxfail) and call.excinfo and isinstance(call.excinfo.value, xfail.Exception):`
  - Change `elif not rep.skipped and xfailed:` to `elif (not item.config.option.runxfail) and not rep.skipped and xfailed:`
- Keep the existing skip-location rewrite block unchanged; once the broad pass is removed, it will run for skip/skipif marks regardless of `--runxfail`.

Reproduction steps and observed failure:
1) Create two tests:

```py
# test_skip.py
import pytest

@pytest.mark.skip(reason="skip me")
def test_skip():
    assert True
```

```py
# test_skipif.py
import pytest

@pytest.mark.skipif(True, reason="conditional skip")
def test_skipif():
    assert True
```

2) Commands and outputs (before fix):
- `pytest -q -rs test_skip.py` → `SKIP [1] test_skip.py:3: skip me` (correct)
- `pytest -q -rs test_skip.py --runxfail` → `SKIP [1] src/_pytest/skipping.py:239: skip me` (incorrect)
- `pytest -q -rs test_skipif.py` → `SKIP [1] test_skipif.py:4: conditional skip` (correct)
- `pytest -q -rs test_skipif.py --runxfail` → `SKIP [1] src/_pytest/skipping.py:239: conditional skip` (incorrect)

3) After applying the fix:
- `pytest -q -rs test_skip.py --runxfail` → `SKIP [1] test_skip.py:3: skip me` (correct)
- `pytest -q -rs test_skipif.py --runxfail` → `SKIP [1] test_skipif.py:4: conditional skip` (correct)

XFail behavior check:

```py
# test_xfail.py
import pytest

@pytest.mark.xfail(reason="expected failure")
def test_xfail_fails():
    assert False
```

- `pytest -q -rs test_xfail.py` → shows XFAIL as usual.
- `pytest -q -rs --runxfail test_xfail.py` → xfail handling disabled; test fails normally without XFAIL/XPASS, as expected.

Edge cases:
- Skips triggered via marks (skip/skipif) will have locations rewritten to the item under `--runxfail`.
- Direct `pytest.skip()` calls inside tests are unaffected (location stays where called in test).
- unittest-style tests and unexpected successes remain unaffected.

We will implement this fix via a PR targeting branch `pytest-dev__pytest-7432`.

### PR DESCRIPTION
**Pull request by casey-brooks**
## Summary
- guard xfail-only report branches when `--runxfail` is active so the skip-location rewrite still executes
- remove the early `--runxfail` bypass to keep skip reports tied to their test definitions

Fixes #29.

## Observed failure (pre-fix)
Commands were run from `/workspace/tmp_skip_tests` with the repo venv activated (`PYTEST_DISABLE_PLUGIN_AUTOLOAD=1`).

```shell
PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 .venv/bin/pytest -q -rs test_skip.py --runxfail
```
```text
s                                                                        [100%]
=========================== short test summary info ============================
SKIPPED [1] ../pytest/src/_pytest/skipping.py:239: skip me
1 skipped in 0.00s
```

```shell
PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 .venv/bin/pytest -q -rs test_skipif.py --runxfail
```
```text
s                                                                        [100%]
=========================== short test summary info ============================
SKIPPED [1] ../pytest/src/_pytest/skipping.py:239: conditional skip
1 skipped in 0.00s
```

## Verification
```shell
PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 .venv/bin/pytest -q -rs test_skip.py --runxfail
```
```text
s                                                                        [100%]
=========================== short test summary info ============================
SKIPPED [1] test_skip.py:4: skip me
1 skipped in 0.00s
```

```shell
PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 .venv/bin/pytest -q -rs test_skipif.py --runxfail
```
```text
s                                                                        [100%]
=========================== short test summary info ============================
SKIPPED [1] test_skipif.py:4: conditional skip
1 skipped in 0.00s
```

```shell
PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 .venv/bin/pytest -q -rs test_xfail.py
```
```text
x                                                                        [100%]
1 xfailed in 0.01s
```

```shell
PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 .venv/bin/pytest -q -rs test_xfail.py --runxfail
```
```text
F                                                                        [100%]
=================================== FAILURES ===================================
_______________________________ test_xfail_fails _______________________________

    @pytest.mark.xfail(reason="expected failure")
    def test_xfail_fails():
>       assert False
E       assert False

test_xfail.py:6: AssertionError
1 failed in 0.01s
```

## Notes
- Direct `pytest.skip` calls continue to respect the original call site.
- `unittest`-style tests still report unexpected successes/failures exactly as before.
- CI intentionally not triggered per repository policy.

### CODE REVIEW

**Comment by casey-brooks** [comment]
### Test & Lint Summary
- `PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 .venv/bin/pytest -q -rs test_skip.py` → 1 skipped
- `PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 .venv/bin/pytest -q -rs test_skip.py --runxfail` → 1 skipped
- `PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 .venv/bin/pytest -q -rs test_skipif.py` → 1 skipped
- `PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 .venv/bin/pytest -q -rs test_skipif.py --runxfail` → 1 skipped
- `PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 .venv/bin/pytest -q -rs test_xfail.py` → 1 xfailed
- `PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 .venv/bin/pytest -q -rs test_xfail.py --runxfail` → 1 failure (expected under `--runxfail`)
- `git diff -- src/_pytest/skipping.py | flake8 --diff` → no lint findings

File: src/_pytest/skipping.py [hunk 1]
```diff
@@ -270,13 +270,15 @@ def pytest_runtest_makereport(item: Item, call: CallInfo[None]):
...
+    elif (
+        not item.config.option.runxfail
+        and call.excinfo
+        and isinstance(call.excinfo.value, xfail.Exception)
+    ):
         assert call.excinfo.value.msg is not None
         rep.wasxfail = "reason: " + call.excinfo.value.msg
         rep.outcome = "skipped"
-    elif not rep.skipped and xfailed:
+    elif (not item.config.option.runxfail) and not rep.skipped and xfailed:
```
Comment (noa-lucent) [inline src/_pytest/skipping.py:281]: [major] Thanks for tightening the guards so that marked skips still get their location rewritten under . Right now this behavior isn\'t exercised by the test suite, though, so the regression could easily slip back in. Please add a test in  that demonstrates the bug (e.g. a  test run with  asserting the reported location is the test file) so we keep coverage on this path.

Comment (noa-lucent) [inline src/_pytest/skipping.py:281]: [major] Reposting with full context: thanks for tightening the guards so that marked skips still get their location rewritten under --runxfail. Right now this behavior isn't covered by the test suite, so the regression could easily slip back in.

Please add a regression test (e.g. in testing/test_skipping.py) that exercises a @pytest.mark.skip or skipif test executed with -rs --runxfail and asserts the reported location is the test file. That will keep this path locked down.

**Review by noa-lucent — CHANGES_REQUESTED** [review]
The fix looks correct, but we need a regression test so this behavior stays covered. Please add coverage for the skip-location rewrite when running with --runxfail.

**Comment by casey-brooks** [comment]
Added regression coverage:
-   ..                                                                       [100%]
2 passed in 0.03s → 2 passed

File: src/_pytest/skipping.py (hunk 1 above)
Comment (noa-lucent) [reply src/_pytest/skipping.py:281]: Added regression coverage in testing/test_skipping.py ensuring --runxfail keeps skip locations tied to the item. Thanks!

Comment (noa-lucent) [reply src/_pytest/skipping.py:281]: Regression tests added in testing/test_skipping.py cover both skip and skipif marks under --runxfail.

**Review by noa-lucent — APPROVED** [review]
Regression coverage looks good—thanks for adding the tests.
